Avoid storing effective language on ElementRareData when it matches the effective document element language

heycam opened the pull request:
#### f31c6cd343763c953418192bbcea0982c167f7f7
<pre>
Avoid storing effective language on ElementRareData when it matches the effective document element language
<a href="https://bugs.webkit.org/show_bug.cgi?id=251657">https://bugs.webkit.org/show_bug.cgi?id=251657</a>
&lt;rdar://problem/104987630&gt;

Reviewed by Ryosuke Niwa.

When an element&apos;s lang attribute is set, we set the effective language on
the ElementRareData of all elements in the subtree. We have an existing
optimization that avoids this if we&apos;re setting it on the document
element. There are some pages, like Wikipedia, that set an explicit
lang on some other element in the body of the page, but which matches
the document element&apos;s lang.

We can avoid the memory overhead of allocating ElementRareData to store
the effective language in this case by using a flag on the element. On
large Wikipedia pages this can save several MB.

If the document element language changes later, we must update the
effective lang state on subtrees that are using this flag. A WeakHashSet
is added to Document to track elements that have an explicit lang
attribute that matches the effective document element language.

Three flags on EventTarget are introduced:

- HasLangAttr and HasXMLLangAttr: records that the element has the
  corresponding lang attribute. This allows us to avoid searching
  for an attribute when updating the effective lang state.

- EffectiveLangKnownToMatchDocument: records that the element has an
  effective lang that matches the effective document element language,
  whether it&apos;s due to an explicit lang attribute or inherited from an
  ancestor. This flag is used in place of
  ElementRareData::m_effectiveLang.

The EffectiveLangKnownToMatchDocument flag is used in place of the &quot;null
effective language means we&apos;ve inherited the effective document
language&quot; state so that disconnected subtrees can also make use of this
optimization. Otherwise, for a case like this:

  let e = document.createElement(&quot;div&quot;);
  e.lang = &quot;en&quot;;  // matching document
  e.append(document.createElement(&quot;div&quot;));

the child element would not know whether to return nullptr or &quot;en&quot; from
Element::effectiveLang() without looking up the tree to see if there is
an ancestor with a langauge attribute.

The EffectiveLangKnownToMatchDocument flag is not
EffectiveLangMatchesDocument, since we don&apos;t set it if the document
element language changes and an existing element starts matching it.
Rather than track all elements with lang attributes to handle such
cases, we leave the effective lang stored on the ElementRareData.

* LayoutTests/fast/css/lang-matching-document-invalidation-expected.txt: Added.
* LayoutTests/fast/css/lang-matching-document-invalidation.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDocumentElementLanguage):
(WebCore::Document::addElementWithLangAttrMatchingDocument):
(WebCore::Document::removeElementWithLangAttrMatchingDocument):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::~Element):
(WebCore::Element::attributeChanged):
(WebCore::Element::setEffectiveLangInSubtree):
(WebCore::Element::didMoveToNewDocument):
(WebCore::Element::setEffectiveLangFromParent):
(WebCore::Element::setEffectiveLang):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::hasEffectiveLangState const):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::effectiveLang const):
(WebCore::Element::langFromAttribute const):
(WebCore::Element::langAttrMatchesDocument const):
(WebCore::Element::setLangAttrMatchesDocument):
(WebCore::Element::effectiveLangMatchesDocument const):
(WebCore::Element::setEffectiveLangMatchesDocument):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/EventTarget.h:

Canonical link: <a href="https://commits.webkit.org/260047@main">https://commits.webkit.org/260047@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6872768c4ffef560add8ce941904a27a793a1933

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116002 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115571 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110721 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7053 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99006 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112586 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96126 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40721 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95017 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27778 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9016 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29133 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6140 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15185 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48677 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11112 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3748 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2650,12 +2651,28 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
     }
 
-    if (!hasEffectiveLangState())
+    if (parentNode() == &parentOfInsertedTree && parentNode()->isDocumentNode()) {
```
A more idiomatic way of checking whether parent is a document is `is<Document>(*parentNode())`.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2718,7 +2735,9 @@ void Element::removedFromAncestor(RemovalType removalType, ContainerNode& oldPar
 
-    if (!hasEffectiveLangState())
+    if (!parentNode() && oldParentOfRemovedTree.isDocumentNode())
+        document().setDocumentElementLanguage(nullAtom());
```
Don't we also want to clear hasLangAttrKnownToMatchDocumentElement flag here as well?

heycam replied:
That flag should never be set on the document element, and an element that is inserted as a new document element will have the flag cleared in Element::clearEffectiveLangStateOnNewDocumentElement.

heycam committed:
Avoid storing effective language on ElementRareData when it matches the effective document element language
https://bugs.webkit.org/show_bug.cgi?id=251657
<rdar://problem/104987630>

Reviewed by Ryosuke Niwa.

When an element's lang attribute is set, we set the effective language on
the ElementRareData of all elements in the subtree. We have an existing
optimization that avoids this if we're setting it on the document
element. There are some pages, like Wikipedia, that set an explicit
lang on some other element in the body of the page, but which matches
the document element's lang.

We can avoid the memory overhead of allocating ElementRareData to store
the effective language in this case by using a flag on the element. On
large Wikipedia pages this can save several MB.

If the document element language changes later, we must update the
effective lang state on subtrees that are using this flag. A WeakHashSet
is added to Document to track elements that have an explicit lang
attribute that matches the effective document element language.

Three flags on EventTarget are introduced:

- HasLangAttr and HasXMLLangAttr: records that the element has the
  corresponding lang attribute. This allows us to avoid searching
  for an attribute when updating the effective lang state.

- EffectiveLangKnownToMatchDocument: records that the element has an
  effective lang that matches the effective document element language,
  whether it's due to an explicit lang attribute or inherited from an
  ancestor. This flag is used in place of
  ElementRareData::m_effectiveLang.

The EffectiveLangKnownToMatchDocument flag is used in place of the "null
effective language means we've inherited the effective document
language" state so that disconnected subtrees can also make use of this
optimization. Otherwise, for a case like this:

  let e = document.createElement("div");
  e.lang = "en";  // matching document
  e.append(document.createElement("div"));

the child element would not know whether to return nullptr or "en" from
Element::effectiveLang() without looking up the tree to see if there is
an ancestor with a langauge attribute.

The EffectiveLangKnownToMatchDocument flag is not
EffectiveLangMatchesDocument, since we don't set it if the document
element language changes and an existing element starts matching it.
Rather than track all elements with lang attributes to handle such
cases, we leave the effective lang stored on the ElementRareData.

* LayoutTests/fast/css/lang-matching-document-invalidation-expected.txt: Added.
* LayoutTests/fast/css/lang-matching-document-invalidation.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDocumentElementLanguage):
(WebCore::Document::addElementWithLangAttrMatchingDocument):
(WebCore::Document::removeElementWithLangAttrMatchingDocument):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::~Element):
(WebCore::Element::attributeChanged):
(WebCore::Element::setEffectiveLangInSubtree):
(WebCore::Element::didMoveToNewDocument):
(WebCore::Element::setEffectiveLangFromParent):
(WebCore::Element::setEffectiveLang):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::hasEffectiveLangState const):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::effectiveLang const):
(WebCore::Element::langFromAttribute const):
(WebCore::Element::langAttrMatchesDocument const):
(WebCore::Element::setLangAttrMatchesDocument):
(WebCore::Element::effectiveLangMatchesDocument const):
(WebCore::Element::setEffectiveLangMatchesDocument):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/EventTarget.h:

Canonical link: https://commits.webkit.org/260047@main
